File: leetcode/simplifyPath.py
# coding: utf-8
__author__ = 'ce39906'
'''
Given an absolute path for a file (Unix-style), simplify it.

For example,
path = "/home/", => "/home"
path = "/a/./b/../../c/", => "/c"
'''
def simplifyPath(path):
    pathArray=path.split('/')
    #print pathArray
    j=0
    while j<len(pathArray):
        if len(pathArray[j])==0:
            del pathArray[j]
        else:
            j+=1
    print pathArray
    if pathArray==[]:
        return '/'
    stack=[]
    for i in pathArray:
        if i!='..' and i!='.':
            stack.append(i)
        elif i=='..':
            if len(stack)!=0:
                stack.pop()
    return '/'+'/'.join(stack)
path='/home/of/foo/../../bar/../../is/./here/.'
print simplifyPath(path)

File: leetcode/simplifyPath.cpp
/*************************************************************************
    > File Name: simplifyPath.cpp
    > Author: ce39906
    > Mail: ce39906@163.com 
    > Created Time: 2018-08-30 19:45:04
 ************************************************************************/
class Solution {
public:
    string simplifyPath(string path) {
        if (path.empty())
        {
            return "";
        }
        // first split path by '/'
        stringstream ss(path);
        vector<string> dirs;
        string dir;
        while (std::getline(ss, dir, '/'))
        {
            if (!dir.empty() && dir != ".")
            {
                dirs.push_back(dir);
            }
        }
        
        stack<string> s;
        for (const auto& dir : dirs)
        {
            if (dir == "..")
            {
                if (!s.empty())
                {
                    s.pop();
                }
            }
            else
            {
                s.emplace(dir);
            }
        }
        
        string res = "";
        while (!s.empty())
        {
            res = s.top() + "/" + res;
            s.pop();
        }
        
        res = "/" + res;
        
        if (res.size() > 1 && res.back() == '/')
        {
            res.resize(res.size() - 1);   
        }
        
        return res;
    }
};
